File: ultrasoundgeneration.py
import itk

from constants import *
import os
import scipy.linalg
print("whoop")

train_volumes_path = volumes_path + "train/"
test_volumes_path = volumes_path + "test/"
train_names = os.listdir(train_volumes_path)
test_names = os.listdir(test_volumes_path)


train_output_dir="prepped_data/train/"
test_output_dir="prepped_data/test/"

#import itkwidgets

import numpy as np
import matplotlib.pyplot as plt

debug=False

import pickle
def load_image_annotation(name, folder):
    image = itk.imread(folder + name)
    annotations = [
        np.argmax(pickle.load(
            open(
                annotations_path + name[:-5] + "axis" + str(i) + ".pickle",
                "rb"
            )
        )[:, :, 3], 0)
        for i in [0, 1]
    ]
    for i in range(2):
        annotations[i] = image.GetSpacing()[i] * annotations[i]
    return image, annotations
class TwoSensorProbe:
    def __init__(self, side_offset, angle):
        self.side_offset = side_offset
        self.angle = angle
    
    def get_origin_direction(self, origin, direction, idx):
        """idx = 1 or -1"""
        origin = np.array(origin) +  idx * self.side_offset * direction[:, 2]
        angle = self.angle * idx
        direction = np.dot(direction, np.array([[1, 0, 0],
                                                [0, np.cos(angle), np.sin(angle)],
                                                [0, -np.sin(angle), np.cos(angle)]]))
        return origin, direction
    
def slice_multiprobe(image, probe, origin, direction):
    res = []
    for idx in [-1, 1]:
        o, d = probe.get_origin_direction(origin, direction, idx)
        if debug:
            pass#print(o, d)
        p = SliceParams(
            o, d,
            width=100,  #mm
            height=100) #mm

        x = spine_slice(image, p)

        x = itk.GetArrayFromImage(x)
        x = x[0]
        res.append(x)

    return res
         

class SliceParams:
    def __init__(self, origin, direction=np.identity(3), width=100, height=100, px=128):
        self.direction=direction
        self.origin=origin
        self.width=width
        self.height=height
        self.px=px
        
        assert len(self.origin) == 3
        
def spine_slice(image, params):
    """
    image is the ITK image to slice from
    params is a SliceParams object"""
    output_spacing = [params.height / params.px, params.width / params.px, 1]
    
    
    output_origin = [
            val + params.direction[j, 0] * -params.height / 2 + params.direction[j, 1] * -params.width / 2
            for j, val in enumerate(params.origin)
        ]
    
    #print(output_origin)
    #print(output_spacing)

    return itk.resample_image_filter(
        image, 
        output_origin = output_origin,
        output_spacing = output_spacing, 
        OutputDirection=itk.matrix_from_array(np.ascontiguousarray(params.direction)), 
        size=[params.px, params.px, 1])
import matplotlib.pyplot as plt


def random_small_rotation(factor = 10):
    R = scipy.linalg.orth(np.random.randn(3, 3).astype(np.float64))
    lam, J = scipy.linalg.eig(R)
    RR = np.real(np.dot(np.dot(J, np.diag(lam)**(1 / factor)), np.linalg.inv(J)))
    return RR

def shitty_ultrasound(y):
    

    y = y + 1000
    y = abs(1 - y[:, 1:] / y[:, :-1])

    denominator = np.cumsum(y, axis=1) + .01
    return np.maximum(-10, np.log(np.abs(y / 4**denominator + .0001 * np.random.randn(128, 128)) ))

def generate_sample(image, annotations, evil_debug=False):
    slice_idx = np.random.randint(50, len(annotations[0]) - 50)

    direction = np.dot(random_small_rotation(), np.array([[0, 0, 1],
                  [-1, 0, 0],
                  [0, -1, 0]], dtype=np.float64))

    origin = np.array([annotations[1][slice_idx], -annotations[0][slice_idx], -slice_idx]) + np.random.randn(3) * 15

    probe = TwoSensorProbe(20, -np.pi /4)


    a, b = slice_multiprobe(image, probe, origin, direction)
    if debug:
        plt.imshow(a)
        plt.show()
        plt.imshow(b)
        plt.show()


    movement_relative_to_image_1 = np.random.randn(3) * 10
    rotation_relative_to_image_1 = random_small_rotation(factor=25)

    direction_2 = np.dot(direction, rotation_relative_to_image_1)
    origin_2 = origin + np.transpose(np.dot(direction, movement_relative_to_image_1.transpose()))

    c, d = slice_multiprobe(image, probe, origin_2, direction_2)
    if debug: 
        plt.imshow(c)
        plt.show()
        plt.imshow(d)
        plt.show()
    
    horror = {}
    if evil_debug:
        horror = {k:v for k, v in locals().items() if not k.startswith('__')}
        
    return {"data": [a, b, c, d], "classes": [movement_relative_to_image_1, rotation_relative_to_image_1], "locals":horror}


def generate_data(volume_path, names):
    res = []
    for name in names:
        #print(name)

        image, annotations = load_image_annotation(name, volume_path)

        

        for _ in range(128):
            res.append(generate_sample(image, annotations))

    return res

import os
def load_dataset(path, simulate_ultrasound=False):
    dataset = generate_data(path, os.listdir(path))

    data = []
    classes = []
    for elem in dataset:
        if not simulate_ultrasound:
            data_entry = np.array(np.stack(elem["data"], axis=-1), dtype=np.float32)
            data_entry += 1000
            data_entry /= 2000
        else: 
            data_entry = [shitty_ultrasound(d) for d in elem["data"]]
            data_entry = np.stack(data_entry, axis=-1) / 10 + 1
            data_entry = np.nan_to_num(data_entry)
            data_entry = np.clip(data_entry, -5, 5)

        
        data.append(data_entry)

        class_entry = elem["classes"]
        class_entry = np.concatenate([class_entry[0] / 4, class_entry[1].flatten() * 40])
        class_entry[[3, 7, 11]] -= 40
        classes.append(class_entry)
    data = np.array(data)
    classes = np.array(classes)
    return data, classes

File: ultrasoundgeneration.jl
using PyCall
itk = pyimport("itk")
np = pyimport("numpy")


using StaticArrays
using CuArrays, CUDAnative
using CuArrays, CuTextures
using NRRD
using Rotations


using LinearAlgebra

Vec3 = SVector{3,Float32}
Mat3 = SMatrix{3, 3, Float32, 9}

struct TwoSensorProbe
    side_offset::Float64
    angle::Float64
end

struct SliceParams #measured from center
    origin::Array{Float64,1}
    direction::Array{Float64, 2}
    width::Float64
    height::Float64
    px::Int64
end

struct ItkImage
    array::Any
    direction::Mat3
    spacing::Vec3
    largestPossibleRegion::Vec3
    ItkImage(pyobject) = new(
        np.array(pyobject), 
        Mat3(itk.array_from_matrix(pyobject.GetDirection())),
        Vec3(np.array(pyobject.GetSpacing())),
        Vec3(np.array(pyobject.GetLargestPossibleRegion().GetSize()))
    )
end

struct GPUItkImage
    texture::Any
    direction::Mat3
    spacing::Vec3
    largestPossibleRegion::Vec3
    GPUItkImage(object) = new(
        CuTexture(CuTextureArray(Float32.(object.array))), 
        object.direction,
        object.spacing,
        object.largestPossibleRegion
    )
end

function get_origin_direction(probe::TwoSensorProbe, origin, direction, idx)
    @assert(idx == 1 || idx == -1)
    @. origin = origin + idx * probe.side_offset * direction[:, 3]
    angle = probe.angle * idx
    direction = direction * [1 0 0; 0 cos(angle) sin(angle); 0 -sin(angle) cos(angle)]
    return origin, direction
end

function random_small_rotation(factor)
    #R, _ = qr(randn((3, 3)))
    #lambda, J = eigen(R)
    #return real(J * Diagonal(lambda .^ (1 / factor)) * inv(J))
    return RotXYZ(pi / factor .* randn(3)...)
end

function generate_sample(jimage::ItkImage, annotation)
    gimage= GPUItkImage(jimage)
    
    res = generate_sample(gimage, annotation)
    
    CuTextures.unsafe_free!(gimage.texture)
    return res
end

function generate_sample(jimage::GPUItkImage, annotation)
    initial_position_randomness_scale = 15 #mm
    movement_scale = 10 #mm
    slice_idx = rand(50:length(annotation[1]) - 50) #vals["slice_idx"]

    direction = random_small_rotation(15) * [0 0 1; -1 0 0; 0 -1 0]

    #direction = vals["direction"]

    # depends on annotation having 1 mm spacing along spine
    origin = (
        [annotation[2][slice_idx], -annotation[1][slice_idx], -slice_idx] 
        .+ initial_position_randomness_scale .* randn(Float64, 3) 
    )

    #debug
    #origin = vals["origin"]

    probe = TwoSensorProbe(20, -np.pi / 4)

    a, b = slice_multiprobe(jimage, probe, origin, direction)

    movement_relative_to_image_1 = randn(3) .* movement_scale
    rotation_relative_to_image_1 = random_small_rotation(25)

    direction_2 = direction * rotation_relative_to_image_1
    origin_2 = origin .+ direction * movement_relative_to_image_1

    c, d = slice_multiprobe(jimage, probe, origin_2, direction_2)

    return Dict(["data"=> [a, b, c, d], "classes"=> [movement_relative_to_image_1, rotation_relative_to_image_1]])
end


function warp(dst, texture, origin::Vec3, direction::Mat3, image_direction::Mat3, image_spacing::Vec3, size::Vec3)
    i::Int32 = (blockIdx().x - 1) * blockDim().x + threadIdx().x
    j::Int32 = (blockIdx().y - 1) * blockDim().y + threadIdx().y
    #u = (Float32(i) - 1f0) / (Float32(size(dst, 1)) - 1f0)
    #v = (Float32(j) - 1f0) / (Float32(size(dst, 2)) - 1f0)
    #x = u  #+ 0.02f0 * CUDAnative.sin(30v)
    #y = v #+ 0.03f0 * CUDAnative.sin(20u)
    
    pixel_position = origin .+ direction[:, 1] .* Float32(j) 
    pixel_position = pixel_position .+ direction[:, 2] .* Float32(i)
    
    pixel_position = pixel_position ./ image_spacing
    direction_flat = (image_direction[1, 1], image_direction[2, 2], image_direction[3, 3])
    pixel_position = pixel_position ./ direction_flat
    
    
    @inbounds dst[i,j] = texture(pixel_position[3] / size[3], pixel_position[2] / size[2], pixel_position[1] / size[1])
    return nothing
end

function spine_slice(image, params)
    output_direction = params.direction * Diagonal([params.height / params.px, params.width / params.px, 0])
    
    output_origin = params.origin .+ params.direction * [-params.height / 2; -params.width / 2; 0]
    
    outimg_d = CuArray{Float32}(undef, 128, 128)
    @cuda threads = (128, 1) blocks = (1, 128) warp(
        outimg_d, image.texture, Vec3(output_origin), Mat3(output_direction), image.direction, image.spacing,
        image.largestPossibleRegion
    )
    return Array(outimg_d)
end

function slice_multiprobe(jimage, probe, origin, direction)
    res = []
    params_array = []
    for idx = [-1, 1]
        slice_origin, slice_direction = get_origin_direction(probe, origin, direction, idx)
        #println(slice_origin)
        #print(slice_direction)
        params = SliceParams(slice_origin, slice_direction, 100, 100, 128)
        x = spine_slice(jimage, params)
        push!(res, x)
        push!(params_array, params)
    end
    return res
end

function generate_data(jimages, annotations)
    res = []#::Array{Dict{String,Array{T,1} where T}, 1} = []
    for(jimage, annotation) = zip(jimages, annotations)
        gimage = GPUItkImage(jimage)
        for(i) = 1:67
            push!(res, generate_sample(gimage, annotation))
            #push!(res, ultrasoundgeneration.generate_sample(
            #        jimage, annotation))
        end
        CuTextures.unsafe_free!(gimage.texture)
    end
    print("i")
    
    data = [] #::Array{Array{Float32, 3}, 1} = []
    classes = []
    for elem = res
        data_entry = cat(
            [reshape(x, Val(3)) for x in elem["data"]]...;
            dims=3
        ).* 1.0f0
        data_entry .+= 1000
        data_entry ./= 2000
        
        push!(data, np.array(data_entry))
        
        c = elem["classes"] 
        class_entry = vcat(c[1] ./ 4, 4 .* [c[2].theta1, c[2].theta2, c[2].theta3])
        #class_entry[[4, 8, 12]] .-= 40
        push!(classes, class_entry)
    end
    return data, classes
end

function longcat(data)
    data2 = zeros(Float32, (length(data), size(data[1])...))
    for i = 1:length(data)
        data2[i, :, :, :] = data[i]
    end
    return data2
end
